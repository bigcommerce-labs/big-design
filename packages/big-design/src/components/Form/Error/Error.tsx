import React from 'react';

import { Small, SmallProps } from '../../Typography';

export const Error: React.FC<SmallProps> = ({ className, style, ...props }) => (
  <Small color="danger" margin="none" marginLeft="xxSmall" {...props} />
);
