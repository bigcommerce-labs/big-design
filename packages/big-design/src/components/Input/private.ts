export { StyledInput, StyledInputWrapper, type StyledInputWrapperProps } from './styled';
