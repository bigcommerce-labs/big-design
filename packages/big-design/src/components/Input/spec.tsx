import { AddIcon } from '@bigcommerce/big-design-icons';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';

import { Form } from '../Form';

import { Input } from './index';

test('forwards ref', () => {
  const ref = React.createRef<HTMLInputElement>();
  const { container } = render(<Input ref={ref} />);
  const input = container.querySelector('input');

  expect(ref.current).toBe(input);
});

test('renders an input tag', () => {
  const { container } = render(<Input />);

  expect(container.querySelector('input')).toBeInTheDocument();
});

test('renders an input with matched label', () => {
  const { queryByLabelText } = render(<Input label="Test Label" />);

  // This one checks for matching id and htmlFor
  expect(queryByLabelText('Test Label')).toBeInTheDocument();
});

test('create unique ids if not provided', () => {
  const { queryByTestId } = render(
    <>
      <Input label="Test Label" data-testid="item1" />
      <Input label="Test Label" data-testid="item2" />
    </>,
  );

  const item1 = queryByTestId('item1') as HTMLInputElement;
  const item2 = queryByTestId('item2') as HTMLInputElement;

  expect(item1).toBeDefined();
  expect(item2).toBeDefined();
  expect(item1.id).not.toBe(item2.id);
});

test('respects provided id', () => {
  const { container } = render(<Input id="test" label="Test Label" />);
  const input = container.querySelector('#test') as HTMLInputElement;

  expect(input.id).toBe('test');
});

test('matches label htmlFor with id provided', () => {
  const { container } = render(<Input id="test" label="Test Label" />);
  const label = container.querySelector('label') as HTMLLabelElement;

  expect(label.htmlFor).toBe('test');
});

test('renders a description', () => {
  const descriptionText = 'This is a description';
  const { queryByText } = render(<Input description={descriptionText} />);

  expect(queryByText(descriptionText)).toBeInTheDocument();
});

test('renders an error', () => {
  const errorText = 'This is an error';
  const { queryByText } = render(
    <Form.Group>
      <Input error={errorText} />
    </Form.Group>,
  );

  expect(queryByText(errorText)).toBeInTheDocument();
});

test('accepts a Label Component', () => {
  const CustomLabel = (
    <Input.Label>
      This is a custom Label
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Label>
  );

  const { queryByTestId } = render(<Input label={CustomLabel} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Label Components', () => {
  const NotALabel = (
    <div>
      This is a not custom Label Component
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Input label={NotALabel} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts a Description Component', () => {
  const CustomDescription = (
    <Input.Description>
      This is a custom Description
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Description>
  );

  const { queryByTestId } = render(<Input description={CustomDescription} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Description Components', () => {
  const NotADescription = (
    <div>
      This is a not custom description
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Input description={NotADescription} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts an Error Component', () => {
  const CustomError = (
    <Input.Error>
      This is a custom Error Component
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Error>
  );

  const { queryByTestId } = render(
    <Form.Group>
      <Input error={CustomError} />
    </Form.Group>,
  );

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Error Components', () => {
  const NotAnError = (
    <div>
      This is a not a custom error component
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(
    <Form.Group>
      <Input error={NotAnError} />
    </Form.Group>,
  );

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('renders iconLeft', () => {
  const { queryByTestId } = render(<Input iconLeft={<AddIcon data-testid="icon" />} />);

  expect(queryByTestId('icon')).toBeInTheDocument();
});

test('renders iconRight', () => {
  const { queryByTestId } = render(<Input iconRight={<AddIcon data-testid="icon" />} />);

  expect(queryByTestId('icon')).toBeInTheDocument();
});

test('renders both icons', () => {
  const { queryByTestId } = render(
    <Input iconRight={<AddIcon data-testid="icon-right" />} iconLeft={<AddIcon data-testid="icon-left" />} />,
  );

  expect(queryByTestId('icon-left')).toBeInTheDocument();
  expect(queryByTestId('icon-right')).toBeInTheDocument();
});

test('renders all together', () => {
  const { container } = render(
    <Input
      label="This is a label"
      description="This is a description"
      iconRight={<AddIcon data-testid="icon-right" />}
      iconLeft={<AddIcon data-testid="icon-left" />}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('is invalid when required and has no value', () => {
  const { getByLabelText } = render(<Input label="Label" description="This is a description" required />);

  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(false);
});

test('is valid when required and has a value', () => {
  const { getByLabelText } = render(
    <Input label="Label" value="Hello world" description="This is a description" required />,
  );

  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(true);
});

test('it is invalid when it was an error', () => {
  const { getByLabelText } = render(
    <Input label="Label" error="Some error" description="This is a description" required />,
  );
  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(false);

  fireEvent.change(input, { target: { value: 'Some new value ' } });

  expect(input.checkValidity()).toEqual(false);
});

test('error shows with valid string', () => {
  const error = 'Error';
  const { container, rerender } = render(
    <Form.Group>
      <Input error="" />
    </Form.Group>,
  );

  expect(container.querySelector('[class*="StyledError"]')).not.toBeInTheDocument();

  rerender(
    <Form.Group>
      <Input error={error} />
    </Form.Group>,
  );

  expect(container.querySelector('[class*="StyledError"]')).toBeInTheDocument();
});
