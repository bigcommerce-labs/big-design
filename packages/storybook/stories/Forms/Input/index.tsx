export * from './Input';
export * from './InputPropTable';
