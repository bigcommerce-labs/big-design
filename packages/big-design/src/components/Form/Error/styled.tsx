import styled from 'styled-components';

import { defaultTheme } from '../../../theme';
import { StyleableSmall } from '../../Typography/private';

export const StyledError = styled(StyleableSmall)`
  color: ${({ theme }) => theme.colors.danger40};
  margin-top: ${({ theme }) => theme.spacing.xxSmall};
`;

StyledError.defaultProps = { theme: defaultTheme };
