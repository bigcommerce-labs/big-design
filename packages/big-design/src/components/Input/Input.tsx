import hoistNonReactStatics from 'hoist-non-react-statics';
import React, { Ref } from 'react';

import { uniqueId } from '../../utils';
import { Chip } from '../Chip';
import { Form } from '../Form';
import { Small } from '../Typography';

import { StyledIconWrapper, StyledInput, StyledInputContent, StyledInputWrapper } from './styled';

interface Props {
  chips?: string[];
  description?: React.ReactChild;
  error?: React.ReactChild | React.ReactChild[];
  iconLeft?: React.ReactChild;
  iconRight?: React.ReactChild;
  label?: React.ReactChild;
  labelId?: string;
  onChipDelete?(chip: string): () => void;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLInputElement>;
}

interface InputState {
  focus: boolean;
}

export type InputProps = Props & React.InputHTMLAttributes<HTMLInputElement>;

class StyleableInput extends React.PureComponent<InputProps & PrivateProps, InputState> {
  static Description = Small;
  static Error = Form.Error;
  static Label = Form.Label;

  readonly state: InputState = {
    focus: false,
  };

  private readonly uniqueId = uniqueId('input_');

  render() {
    const { chips, description, disabled, error, label, labelId, forwardedRef, onChipDelete, ...props } = this.props;
    const id = this.getId();

    return (
      <div>
        {this.renderLabel()}
        {this.renderDescription()}
        <StyledInputWrapper disabled={disabled} error={error} focus={this.state.focus}>
          {this.renderIconLeft()}
          <StyledInputContent chips={chips}>
            {this.renderChips()}
            <StyledInput
              {...props}
              disabled={disabled}
              chips={chips}
              error={error}
              id={id}
              onBlur={this.onInputBlur}
              onFocus={this.onInputFocus}
              ref={forwardedRef}
            />
          </StyledInputContent>

          {this.renderIconRight()}
        </StyledInputWrapper>
      </div>
    );
  }

  private getId() {
    const { id } = this.props;

    return id ? id : this.uniqueId;
  }

  private onInputFocus = (event: React.FocusEvent<HTMLInputElement>) => {
    const { onFocus } = this.props;

    this.setFocus(true);

    return onFocus && onFocus(event);
  };

  private onInputBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    const { onBlur } = this.props;

    this.setFocus(false);

    return onBlur && onBlur(event);
  };

  private renderDescription() {
    const { description } = this.props;

    if (typeof description === 'string') {
      return <Input.Description>{description}</Input.Description>;
    }

    if (React.isValidElement(description) && description.type === Input.Description) {
      return description;
    }

    return null;
  }

  private renderLabel() {
    const { label, labelId, required } = this.props;
    const id = this.getId();

    if (typeof label === 'string') {
      return (
        <Input.Label id={labelId} htmlFor={id} renderOptional={!required}>
          {label}
        </Input.Label>
      );
    }

    if (React.isValidElement(label) && label.type === Input.Label) {
      return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
        id: labelId,
        htmlFor: id,
      });
    }

    return null;
  }

  private renderIconLeft() {
    if (!this.props.iconLeft) {
      return null;
    }

    return <StyledIconWrapper>{this.props.iconLeft}</StyledIconWrapper>;
  }

  private renderIconRight() {
    if (!this.props.iconRight) {
      return null;
    }

    return <StyledIconWrapper>{this.props.iconRight}</StyledIconWrapper>;
  }

  private setFocus(toggle: boolean) {
    this.setState({ focus: toggle });
  }

  private renderChips() {
    const { chips, onChipDelete } = this.props;

    if (!chips) {
      return null;
    }

    return chips.map((chip, key) =>
      onChipDelete ? (
        <Chip key={key} label={chip} marginBottom="none" onDelete={onChipDelete(chip)} />
      ) : (
        <Chip key={key} label={chip} marginBottom="none" marginTop="none" />
      ),
    );
  }
}

const InputWithForwardedRef = React.forwardRef<HTMLInputElement, InputProps>(({ className, style, ...props }, ref) => (
  <StyleableInput {...props} forwardedRef={ref} />
));

export const Input = hoistNonReactStatics(InputWithForwardedRef, StyleableInput);

Input.displayName = 'Input';
