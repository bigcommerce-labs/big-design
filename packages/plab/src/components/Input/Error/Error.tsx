import React from 'react';

import { StyledError } from './styled';

export const Error: React.SFC<{}> = props => <StyledError {...props} />;
