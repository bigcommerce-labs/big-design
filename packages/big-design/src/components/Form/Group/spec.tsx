import { render } from '@testing-library/react';
import React from 'react';
import 'jest-styled-components';

import { Input } from '../../Input';
import { FormControlError } from '../Error';

import { FormGroup } from './';

test('renders a form group', () => {
  const { container } = render(<FormGroup />);

  expect(container.firstChild).toMatchSnapshot();
});

test('renders group with input', () => {
  const { container } = render(
    <FormGroup>
      <Input />
    </FormGroup>,
  );

  expect(container.querySelector('input')).toBeInTheDocument();
});

test('renders group and input with error', () => {
  const error = 'Error';
  const { getByText } = render(
    <FormGroup>
      <Input error={error} />
    </FormGroup>,
  );

  expect(getByText(error)).toBeInTheDocument();
});

test('renders group and nested input with error', () => {
  const error = 'Error';
  const { getByText } = render(
    <FormGroup>
      <div>
        <Input error={error} />
      </div>
    </FormGroup>,
  );

  expect(getByText(error)).toBeInTheDocument();
});

test('renders group with error prop', () => {
  const error = 'Error';
  const { getByText } = render(
    <FormGroup errors={error}>
      <Input />
    </FormGroup>,
  );

  expect(getByText(error)).toBeInTheDocument();
});

test('renders error prop with an array of errors', () => {
  const errors = ['Error 1', 'Error 2', 'Error 3'];
  const { getByText } = render(
    <FormGroup errors={errors}>
      <Input />
    </FormGroup>,
  );

  errors.forEach((error) => expect(getByText(error)).toBeInTheDocument());
});

test('renders error with FormControlError element', () => {
  const testId = 'test';
  const errors = <FormControlError data-testid={testId}>Error</FormControlError>;
  const { getByTestId } = render(
    <FormGroup errors={errors}>
      <Input />
    </FormGroup>,
  );

  expect(getByTestId(testId)).toBeInTheDocument();
});

test('renders error prop with an array of FormControlError elements', () => {
  const testIds = ['test_1', 'test_2', 'test_3'];
  const errors = testIds.map((id) => (
    <FormControlError data-testid={id} key={id}>
      Error
    </FormControlError>
  ));
  const { getByTestId } = render(
    <FormGroup errors={errors}>
      <Input />
    </FormGroup>,
  );

  testIds.forEach((id) => expect(getByTestId(id)).toBeInTheDocument());
});

test('does not render invalid errors', () => {
  const testId = 'test';
  const errors = [
    'Error',
    <FormControlError key="1">Error</FormControlError>,
    <div data-testid="testId" key="2">
      Error
    </div>,
  ];
  const { queryByTestId } = render(
    <FormGroup errors={errors}>
      <Input />
    </FormGroup>,
  );

  expect(queryByTestId(testId)).not.toBeInTheDocument();
});
