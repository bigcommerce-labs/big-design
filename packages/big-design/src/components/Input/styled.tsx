import { addValues, theme as defaultTheme, remCalc } from '@bigcommerce/big-design-theme';
import styled, { css } from 'styled-components';

import { PaddingProps, withPaddings } from '../../mixins';
import { withTransition } from '../../mixins/transitions';

import { InputProps } from './Input';

export interface StyledInputWrapperProps extends InputProps {
  focus: boolean;
}

export const StyledInputWrapper = styled.span<StyledInputWrapperProps>`
  ${withTransition(['border', 'box-shadow'])}

  align-items: center;
  background-color: ${({ theme }) => theme.colors.white};
  border-radius: ${({ theme }) => theme.borderRadius.normal};
  box-sizing: border-box;
  display: flex;
  min-height: ${({ theme }) => addValues(theme.spacing.xxLarge, theme.spacing.xxSmall)};
  overflow: hidden;
  position: relative;
  width: 100%;

  ${({ error, theme }) => css`
    border: ${error ? theme.border.boxError : theme.border.box};
  `};

  &:hover:not([disabled]) {
    ${({ error, theme }) =>
      error
        ? css`
            border: ${theme.border.boxError};
          `
        : css`
            border: 1px solid ${theme.colors.secondary40};
          `}
  }

  ${({ error, focus, theme }) =>
    focus &&
    css`
      outline: none;
      box-shadow: 0 0 0 4px ${error ? theme.colors.danger20 : theme.colors.primary20};
    `};

  &[disabled] {
    background-color: ${({ theme }) => theme.colors.secondary20};
  }
`;

export const StyledInput = styled.input<InputProps>`
  background-color: inherit;
  border: 0;
  box-sizing: border-box;
  color: ${({ theme }) => theme.colors.secondary70};
  flex: 1 1 40%;
  height: 100%;
  padding: 0;
  padding-left: ${({ theme }) => theme.spacing.xSmall};
  width: 100%;

  &:focus {
    outline: none;
  }

  &::placeholder {
    color: ${({ theme }) => theme.colors.secondary40};
    font-size: ${({ theme }) => theme.typography.fontSize.medium};
  }

  &:-webkit-autofill {
    /* !important overrides user agent style sheets that use !important in their :-webkit-autofill implementation */
    /* See https://developer.mozilla.org/en-US/docs/Web/CSS/:-webkit-autofill */
    background-color: ${({ theme }) => theme.colors.primary10} !important;
    -webkit-box-shadow: 0 0 0px 1000px ${({ theme }) => theme.colors.primary10} inset;
  }

  ${({ iconRight, theme }) =>
    iconRight &&
    css`
      padding-right: ${addValues(theme.spacing.xxSmall, theme.spacing.xxLarge)};
    `};

  ${({ iconLeft, theme }) =>
    iconLeft &&
    css`
      padding-left: ${addValues(theme.spacing.xxSmall, theme.spacing.xxLarge)};
    `};

  ${({ chips, theme }) =>
    chips &&
    css`
      min-height: ${theme.spacing.xLarge};
      padding-left: ${theme.spacing.xxSmall};
      padding-right: ${theme.spacing.none};
    `};

  ${({ chips, theme }) =>
    chips &&
    chips.length &&
    css`
      margin-top: ${theme.spacing.xxSmall};
    `};

  ${({ chips }) =>
    !chips &&
    css`
      min-height: ${remCalc(34)};
    `};

  &[disabled] {
    background-color: ${({ theme }) => theme.colors.secondary20};
    color: ${({ theme }) => theme.colors.secondary60};
  }
`;

export const StyledIconWrapper = styled.div<PaddingProps>`
  align-items: center;
  color: ${({ theme }) => theme.colors.secondary60};
  display: flex;
  height: 100%;
  position: absolute;
  top: 0;

  ${withPaddings()}

  ${({ paddingLeft }) =>
    paddingLeft === 'xSmall' &&
    css`
      left: 0;
    `}

  ${({ paddingRight }) =>
    paddingRight === 'xSmall' &&
    css`
      right: 0;
    `}
`;

export const StyledInputContent = styled.div<InputProps>`
  align-items: center;
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
  flex: 1;
  height: 100%;

  ${({ chips, theme }) =>
    chips &&
    css`
      margin-left: ${theme.spacing.xxSmall};
      padding-right: ${addValues(theme.spacing.xxSmall, theme.spacing.xxLarge)};
    `};

  ${({ chips, theme }) =>
    chips &&
    chips.length &&
    css`
      margin-bottom: ${theme.spacing.xxSmall};
    `};
`;

StyledInput.defaultProps = { theme: defaultTheme };
StyledInputWrapper.defaultProps = { theme: defaultTheme };
StyledIconWrapper.defaultProps = { theme: defaultTheme };
StyledInputContent.defaultProps = { theme: defaultTheme };
