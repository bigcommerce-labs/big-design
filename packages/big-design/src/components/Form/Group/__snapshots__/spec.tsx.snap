// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a form group 1`] = `
.c0 {
  display: grid;
  grid-gap: 0.5rem 1rem;
  margin-bottom: 1rem;
}

.c0:last-child {
  margin-bottom: 0;
}

@media (min-width:720px) {
  .c0 .styled__StyledInputWrapper-g32raa-0,
  .c0 .styled__StyledTextareaWrapper-c1uos0-0 {
    max-width: 26rem;
  }
}

<div
  class="c0"
/>
`;
